tests/mocha/procedure_map_test.js

```diff
@@ -200,8 +201,494 @@ suite('Procedure Map', function() {
+          procedureModel.insertParameter(
+              0,
+              new Blockly.procedures.ObservableParameterModel(
+                  this.workspace, 'test name'));
+
+          assertEventNotFired(
+            this.eventSpy,
+            Blockly.Events.ProcedureParameterCreate,
+            {},
+            this.workspace.id);
+        });
+
+    test(
+        'parameter delete events are fired when a parameter is deleted',
+        function() {
+          const procedureModel =
+              new Blockly.procedures.ObservableProcedureModel(this.workspace);
+          this.procedureMap.add(procedureModel);
+          const parameterModel =
+              new Blockly.procedures.ObservableParameterModel(
+                  this.workspace, 'test name');
+          procedureModel.insertParameter(0, parameterModel);
+          procedureModel.deleteParameter(0);
+
+          assertEventFired(
+            this.eventSpy,
+            Blockly.Events.ProcedureParameterDelete,
+            {
+              model: procedureModel,
+              parameter: parameterModel,
+              index: 0,
+            },
+            this.workspace.id);
+        });
+
+    test(
+        'parameter delete events are not fired if the procedure is ' +
+        'not in the map',
+        function() {
+          const procedureModel =
+              new Blockly.procedures.ObservableProcedureModel(this.workspace)
+                  .insertParameter(
+                      0,
+                      new Blockly.procedures.ObservableParameterModel(
+                          this.workspace, 'test name'));
+          procedureModel.deleteParameter(0);
+
+          assertEventNotFired(
+            this.eventSpy,
+            Blockly.Events.ProcedureParameterDelete,
+            {},
+            this.workspace.id);
+        });
+
+    test(
+        'parameter rename events are fired when a parameter is renamed',
+        function() {
+          const procedureModel =
+              new Blockly.procedures.ObservableProcedureModel(this.workspace);
+          this.procedureMap.add(procedureModel);
+          const parameterModel =
+              new Blockly.procedures.ObservableParameterModel(
+                  this.workspace, 'test name');
+          procedureModel.insertParameter(0, parameterModel);
+
+          parameterModel.setName('new name');
+
+          assertEventFired(
+            this.eventSpy,
+            Blockly.Events.ProcedureParameterRename,
+            {
+              model: procedureModel,
+              parameter: parameterModel,
+              oldName: 'test name',
+            },
+            this.workspace.id);
+        });
+
+    test(
+        'parameter rename events are not fired if the rename is noop',
+        function() {
+          const procedureModel =
+              new Blockly.procedures.ObservableProcedureModel(this.workspace);
+          this.procedureMap.add(procedureModel);
+          const parameterModel =
```

Comment:
You could pull creating the parameter model out into setup and make this a suite for parameters.

Reply:
I could, but idk if that will make it more readable or less readable. I'll play around with it and see how it feels.